// @mezo-org/orangekit-contracts/typechain/factories/contracts/LegacyERC1271__factory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */

import { Contract, Interface, type ContractRunner } from "ethers";
import type {
  LegacyERC1271,
  LegacyERC1271Interface,
} from "../../contracts/LegacyERC1271";

const _abi = [
  {
    inputs: [
      {
        internalType: "bytes",
        name: "_data",
        type: "bytes",
      },
      {
        internalType: "bytes",
        name: "_signature",
        type: "bytes",
      },
    ],
    name: "isValidSignature",
    outputs: [
      {
        internalType: "bytes4",
        name: "",
        type: "bytes4",
      },
    ],
    stateMutability: "view",
    type: "function",
  },
] as const;

export class LegacyERC1271__factory {
  static readonly abi = _abi;
  static createInterface(): LegacyERC1271Interface {
    return new Interface(_abi) as LegacyERC1271Interface;
  }
  static connect(
    address: string,
    runner?: ContractRunner | null
  ): LegacyERC1271 {
    return new Contract(address, _abi, runner) as unknown as LegacyERC1271;
  }
}
