// @mezo-org/orangekit-smart-account/src/lib/bitcoin/tests/data.ts
// Test data generated using this site: https://www.btcschools.net/bitcoin/
const testData = {
  p2wpkhAddresses: [
    {
      type: "P2WPKH",
      address: "bc1qr2gzwarzhv2r2r70v7522w67rt8n6azcklup4z",
      publicKey:
        "0x0301c3a0b3e1a18d0e04a455e2fbed97d97352cfc35ad272cdc8dcb0fc8dee6707",
      compression: "compressed",
    },
    {
      type: "P2WPKH",
      address: "bc1qxg2cc7p5yur7tr3rr2cl5jw47lluuvy7cr2usu",
      publicKey:
        "0x045fe0ef7394fb466b1bfae8801f5b4d2dfdf01741642dc776d021fd3122618110dc91fe54b9413e9c2b7d92003cf70f7d9ae92a786514646bf138eb0bcdd799c5",
      compression: "uncompressed",
    },
    {
      type: "P2WPKH",
      address: "tb1qhp6ug4hp3cc84spx3l8sddh8w5ffjujvjycqv5",
      publicKey:
        "0x03470e65b45464f6945d7a4434cec1d1df61700f21bf0520f0171cfdab313eeaae",
      compression: "compressed",
    },
    {
      type: "P2WPKH",
      address: "tb1q8xuccgd7juqzy3ualx0c5aqltfqpyt9gmq4yf8",
      publicKey:
        "0x04b4c963521478d89114a33a1b6cc19f8b5db3c35904b6c04a7d3bfb396f21219b3ea38cb306c32550e3e4bace5cd761ca4e845b2fe5a480d654ff224562aea956",
      compression: "uncompressed",
    },
  ],
  p2pkhAddresses: [
    {
      type: "P2PKH",
      address: "15db8HU8AhNaZpX2efV5H5fvJcqmLEgxBB",
      publicKey:
        "0x03d8d91c7e4eb085e2552aa925607f4b622d703df66ab415d916ad90dac1d44369",
      compression: "compressed",
    },
    {
      type: "P2PKH",
      address: "17Nb9dw9yFUPfQTzbFjxk4PzfHqepF3BRF",
      publicKey:
        "0x04811b8ba90b12d4db6066dd8cdadb71bda593d1164cd68e909b9a457b946da0fa21750060c9305ccf847918a983301e355d4c3a5927ebb0361def785ef7587efb",
      compression: "uncompressed",
    },
    {
      type: "P2PKH",
      address: "mpPNkDrN1jReLUpak261U9Mdobk5GZP7U8",
      publicKey:
        "0x0271bbca50d36cf169daecb18ad565de1c4be9231575df57f730ec95d3d86b409f",
      compression: "compressed",
    },
    {
      type: "P2PKH",
      address: "n4kefpTSNgs19nr6zfRDwJA1rZwuzZypSE",
      publicKey:
        "0x04b0386599a1bb687fad4c8995c410fe8f85cba4e55b1e9c7a70db3781a182bc94f4a93b0b580d425c7a879f22120dc26cfab6089ba1c43a3bbf3452421b159593",
      compression: "uncompressed",
    },
  ],
  nestedSegwitAddresses: [
    {
      type: "P2SH-P2WPKH",
      address: "3PqcLLkJ2CDrCXpNbfdmkkPN8PUibE9EVG",
      publicKey:
        "0x0385a365b47f35aff49fdf53cd0b31ab25dec1ef7a4ee3c9d3060ff298ba448d71",
      compression: "compressed",
    },

    {
      type: "P2SH-P2WPKH",
      address: "2N1gqKvdCoFyCpUxsF6HuP2DunnEQRnfuuU",
      publicKey:
        "0x03246de9b5fd955ceba6bcc90d85697b1a74ff8230eef6a1b1a1169ae53b80695d",
      compression: "compressed",
    },
  ],
  unsupportedP2TRAddresses: [
    {
      type: "P2TR",
      address: "bc1p5d7rjq7g6rdk2yhzks9smlaqtedr4dekq08ge8ztwac72sfr9rusxg3297",
    },
  ],
  unsupportedP2SHAddresses: [
    {
      type: "P2SH",
      address: "3JPkm6Cj9vuq5ZxbhV1K6u5tSjSqpLHjdW",
    },
    {
      type: "P2SH",
      address: "2Mygoqi18hjqE8ZyPp7RAJRZJ6SgtsmGFi7",
    },
    {
      type: "P2SH-P2WSH",
      address: "3J2jKQrdbxCL7PRZigwv9MxFEwfamjeH3v",
    },
    {
      type: "P2SH-P2WSH",
      address: "2N1QriLrZFmeVDm5gQCar9ewHLkEdaPywEQ",
    },
    {
      type: "P2SH-LOCKTIME",
      address: "3A4KfvPDYNqRKPgy2LtK6mYi776QCRHdGc",
    },
    {
      type: "P2SH-LOCKTIME",
      address: "2NCEYipTH1MZqQMK4ia8JeV8G2nmMu3442n",
    },
    {
      type: "P2WSH",
      address: "bc1qd73n4nvh3rxgde50h2n3dns2pvfxd20xcrz4u4mng2zt22xqw6rs8mqlfm",
    },
    {
      type: "P2WSH",
      address: "tb1qp0w9cya83zpsekzfrcckjaka8juja83mhl9h7jde3zvf20tfs48qmmsk9w",
    },
  ],
}

export default testData
