// @mezo-org/orangekit-contracts/typechain/contracts/v1.0.0-beta.3/MigrationTo_1_0_0_beta_3.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  EventFragment,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedLogDescription,
  TypedListener,
  TypedContractMethod,
} from "../../common";

export interface MigrationTo_1_0_0_beta_3Interface extends Interface {
  getFunction(
    nameOrSignature:
      | "BITCOIN_SAFE_OWNER_SINGLETON"
      | "EMERGENCY_UPGRADER"
      | "ORANGE_KIT_SAFE_FACTORY_PROXY"
      | "ORANGE_KIT_SAFE_FACTORY_SINGLETON"
      | "SALT"
      | "claimOwnershipBack"
      | "deploy"
      | "deployBitcoinSafeOwnerSingleton"
      | "deployOrangeKitSafeFactorySingleton"
      | "migrate"
  ): FunctionFragment;

  getEvent(
    nameOrSignatureOrTopic:
      | "BitcoinSafeOwnerSingletonDeployed"
      | "OrangeKitSafeFactorySingletonDeployed"
  ): EventFragment;

  encodeFunctionData(
    functionFragment: "BITCOIN_SAFE_OWNER_SINGLETON",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "EMERGENCY_UPGRADER",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "ORANGE_KIT_SAFE_FACTORY_PROXY",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "ORANGE_KIT_SAFE_FACTORY_SINGLETON",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "SALT", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "claimOwnershipBack",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "deploy", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "deployBitcoinSafeOwnerSingleton",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "deployOrangeKitSafeFactorySingleton",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "migrate", values?: undefined): string;

  decodeFunctionResult(
    functionFragment: "BITCOIN_SAFE_OWNER_SINGLETON",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "EMERGENCY_UPGRADER",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "ORANGE_KIT_SAFE_FACTORY_PROXY",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "ORANGE_KIT_SAFE_FACTORY_SINGLETON",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "SALT", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "claimOwnershipBack",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "deploy", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "deployBitcoinSafeOwnerSingleton",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "deployOrangeKitSafeFactorySingleton",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "migrate", data: BytesLike): Result;
}

export namespace BitcoinSafeOwnerSingletonDeployedEvent {
  export type InputTuple = [bitcoinSafeOwnerSingleton: AddressLike];
  export type OutputTuple = [bitcoinSafeOwnerSingleton: string];
  export interface OutputObject {
    bitcoinSafeOwnerSingleton: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace OrangeKitSafeFactorySingletonDeployedEvent {
  export type InputTuple = [orangeKitSafeFactorySingleton: AddressLike];
  export type OutputTuple = [orangeKitSafeFactorySingleton: string];
  export interface OutputObject {
    orangeKitSafeFactorySingleton: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export interface MigrationTo_1_0_0_beta_3 extends BaseContract {
  connect(runner?: ContractRunner | null): MigrationTo_1_0_0_beta_3;
  waitForDeployment(): Promise<this>;

  interface: MigrationTo_1_0_0_beta_3Interface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  BITCOIN_SAFE_OWNER_SINGLETON: TypedContractMethod<[], [string], "view">;

  EMERGENCY_UPGRADER: TypedContractMethod<[], [string], "view">;

  ORANGE_KIT_SAFE_FACTORY_PROXY: TypedContractMethod<[], [string], "view">;

  ORANGE_KIT_SAFE_FACTORY_SINGLETON: TypedContractMethod<[], [string], "view">;

  SALT: TypedContractMethod<[], [string], "view">;

  claimOwnershipBack: TypedContractMethod<[], [void], "nonpayable">;

  deploy: TypedContractMethod<[], [void], "nonpayable">;

  deployBitcoinSafeOwnerSingleton: TypedContractMethod<
    [],
    [void],
    "nonpayable"
  >;

  deployOrangeKitSafeFactorySingleton: TypedContractMethod<
    [],
    [void],
    "nonpayable"
  >;

  migrate: TypedContractMethod<[], [void], "nonpayable">;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "BITCOIN_SAFE_OWNER_SINGLETON"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "EMERGENCY_UPGRADER"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "ORANGE_KIT_SAFE_FACTORY_PROXY"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "ORANGE_KIT_SAFE_FACTORY_SINGLETON"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "SALT"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "claimOwnershipBack"
  ): TypedContractMethod<[], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "deploy"
  ): TypedContractMethod<[], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "deployBitcoinSafeOwnerSingleton"
  ): TypedContractMethod<[], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "deployOrangeKitSafeFactorySingleton"
  ): TypedContractMethod<[], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "migrate"
  ): TypedContractMethod<[], [void], "nonpayable">;

  getEvent(
    key: "BitcoinSafeOwnerSingletonDeployed"
  ): TypedContractEvent<
    BitcoinSafeOwnerSingletonDeployedEvent.InputTuple,
    BitcoinSafeOwnerSingletonDeployedEvent.OutputTuple,
    BitcoinSafeOwnerSingletonDeployedEvent.OutputObject
  >;
  getEvent(
    key: "OrangeKitSafeFactorySingletonDeployed"
  ): TypedContractEvent<
    OrangeKitSafeFactorySingletonDeployedEvent.InputTuple,
    OrangeKitSafeFactorySingletonDeployedEvent.OutputTuple,
    OrangeKitSafeFactorySingletonDeployedEvent.OutputObject
  >;

  filters: {
    "BitcoinSafeOwnerSingletonDeployed(address)": TypedContractEvent<
      BitcoinSafeOwnerSingletonDeployedEvent.InputTuple,
      BitcoinSafeOwnerSingletonDeployedEvent.OutputTuple,
      BitcoinSafeOwnerSingletonDeployedEvent.OutputObject
    >;
    BitcoinSafeOwnerSingletonDeployed: TypedContractEvent<
      BitcoinSafeOwnerSingletonDeployedEvent.InputTuple,
      BitcoinSafeOwnerSingletonDeployedEvent.OutputTuple,
      BitcoinSafeOwnerSingletonDeployedEvent.OutputObject
    >;

    "OrangeKitSafeFactorySingletonDeployed(address)": TypedContractEvent<
      OrangeKitSafeFactorySingletonDeployedEvent.InputTuple,
      OrangeKitSafeFactorySingletonDeployedEvent.OutputTuple,
      OrangeKitSafeFactorySingletonDeployedEvent.OutputObject
    >;
    OrangeKitSafeFactorySingletonDeployed: TypedContractEvent<
      OrangeKitSafeFactorySingletonDeployedEvent.InputTuple,
      OrangeKitSafeFactorySingletonDeployedEvent.OutputTuple,
      OrangeKitSafeFactorySingletonDeployedEvent.OutputObject
    >;
  };
}
