// @mezo-org/orangekit-contracts/typechain/contracts/v1.0.0-beta.3/OrangeKitSafeFactory.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  EventFragment,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedLogDescription,
  TypedListener,
  TypedContractMethod,
} from "../../common";

export interface OrangeKitSafeFactoryInterface extends Interface {
  getFunction(
    nameOrSignature:
      | "bitcoinSafeOwnerProxyCreationCode"
      | "deploySafe"
      | "emergencyGovernance"
      | "initialize"
      | "initialized"
      | "owner"
      | "predictAddresses"
      | "safeOwnerSingleton"
      | "safeProxyCreationCode"
      | "safeSingleton"
      | "transferOwnership"
      | "upgradeSafeOwnerSingleton"
      | "upgradeSingleton"
  ): FunctionFragment;

  getEvent(
    nameOrSignatureOrTopic:
      | "OwnershipTransferred"
      | "SafeDeployed"
      | "SafeOwnerSingletonUpgraded"
      | "SingletonUpgraded"
  ): EventFragment;

  encodeFunctionData(
    functionFragment: "bitcoinSafeOwnerProxyCreationCode",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "deploySafe",
    values: [BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "emergencyGovernance",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "initialize",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "initialized",
    values?: undefined
  ): string;
  encodeFunctionData(functionFragment: "owner", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "predictAddresses",
    values: [BytesLike]
  ): string;
  encodeFunctionData(
    functionFragment: "safeOwnerSingleton",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "safeProxyCreationCode",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "safeSingleton",
    values?: undefined
  ): string;
  encodeFunctionData(
    functionFragment: "transferOwnership",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "upgradeSafeOwnerSingleton",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "upgradeSingleton",
    values: [AddressLike, BytesLike]
  ): string;

  decodeFunctionResult(
    functionFragment: "bitcoinSafeOwnerProxyCreationCode",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "deploySafe", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "emergencyGovernance",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "initialize", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "initialized",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "owner", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "predictAddresses",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "safeOwnerSingleton",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "safeProxyCreationCode",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "safeSingleton",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "transferOwnership",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "upgradeSafeOwnerSingleton",
    data: BytesLike
  ): Result;
  decodeFunctionResult(
    functionFragment: "upgradeSingleton",
    data: BytesLike
  ): Result;
}

export namespace OwnershipTransferredEvent {
  export type InputTuple = [previousOwner: AddressLike, newOwner: AddressLike];
  export type OutputTuple = [previousOwner: string, newOwner: string];
  export interface OutputObject {
    previousOwner: string;
    newOwner: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace SafeDeployedEvent {
  export type InputTuple = [
    safe: AddressLike,
    bitcoinSafeOwner: AddressLike,
    truncatedBitcoinAddress: BytesLike
  ];
  export type OutputTuple = [
    safe: string,
    bitcoinSafeOwner: string,
    truncatedBitcoinAddress: string
  ];
  export interface OutputObject {
    safe: string;
    bitcoinSafeOwner: string;
    truncatedBitcoinAddress: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace SafeOwnerSingletonUpgradedEvent {
  export type InputTuple = [
    oldSingleton: AddressLike,
    newSingleton: AddressLike
  ];
  export type OutputTuple = [oldSingleton: string, newSingleton: string];
  export interface OutputObject {
    oldSingleton: string;
    newSingleton: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export namespace SingletonUpgradedEvent {
  export type InputTuple = [
    oldSingleton: AddressLike,
    newSingleton: AddressLike
  ];
  export type OutputTuple = [oldSingleton: string, newSingleton: string];
  export interface OutputObject {
    oldSingleton: string;
    newSingleton: string;
  }
  export type Event = TypedContractEvent<InputTuple, OutputTuple, OutputObject>;
  export type Filter = TypedDeferredTopicFilter<Event>;
  export type Log = TypedEventLog<Event>;
  export type LogDescription = TypedLogDescription<Event>;
}

export interface OrangeKitSafeFactory extends BaseContract {
  connect(runner?: ContractRunner | null): OrangeKitSafeFactory;
  waitForDeployment(): Promise<this>;

  interface: OrangeKitSafeFactoryInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  bitcoinSafeOwnerProxyCreationCode: TypedContractMethod<[], [string], "view">;

  deploySafe: TypedContractMethod<
    [truncatedBitcoinAddress: BytesLike],
    [[string, string] & { safe: string; bitcoinSafeOwner: string }],
    "nonpayable"
  >;

  emergencyGovernance: TypedContractMethod<[], [string], "view">;

  initialize: TypedContractMethod<[], [void], "nonpayable">;

  initialized: TypedContractMethod<[], [boolean], "view">;

  owner: TypedContractMethod<[], [string], "view">;

  predictAddresses: TypedContractMethod<
    [truncatedBitcoinAddress: BytesLike],
    [[string, string] & { safe: string; bitcoinSafeOwner: string }],
    "view"
  >;

  safeOwnerSingleton: TypedContractMethod<[], [string], "view">;

  safeProxyCreationCode: TypedContractMethod<[], [string], "view">;

  safeSingleton: TypedContractMethod<[], [string], "view">;

  transferOwnership: TypedContractMethod<
    [newOwner: AddressLike],
    [void],
    "nonpayable"
  >;

  upgradeSafeOwnerSingleton: TypedContractMethod<
    [newSafeOwnerSingleton: AddressLike],
    [void],
    "nonpayable"
  >;

  upgradeSingleton: TypedContractMethod<
    [newSingleton: AddressLike, initData: BytesLike],
    [void],
    "nonpayable"
  >;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "bitcoinSafeOwnerProxyCreationCode"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "deploySafe"
  ): TypedContractMethod<
    [truncatedBitcoinAddress: BytesLike],
    [[string, string] & { safe: string; bitcoinSafeOwner: string }],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "emergencyGovernance"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "initialize"
  ): TypedContractMethod<[], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "initialized"
  ): TypedContractMethod<[], [boolean], "view">;
  getFunction(
    nameOrSignature: "owner"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "predictAddresses"
  ): TypedContractMethod<
    [truncatedBitcoinAddress: BytesLike],
    [[string, string] & { safe: string; bitcoinSafeOwner: string }],
    "view"
  >;
  getFunction(
    nameOrSignature: "safeOwnerSingleton"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "safeProxyCreationCode"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "safeSingleton"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "transferOwnership"
  ): TypedContractMethod<[newOwner: AddressLike], [void], "nonpayable">;
  getFunction(
    nameOrSignature: "upgradeSafeOwnerSingleton"
  ): TypedContractMethod<
    [newSafeOwnerSingleton: AddressLike],
    [void],
    "nonpayable"
  >;
  getFunction(
    nameOrSignature: "upgradeSingleton"
  ): TypedContractMethod<
    [newSingleton: AddressLike, initData: BytesLike],
    [void],
    "nonpayable"
  >;

  getEvent(
    key: "OwnershipTransferred"
  ): TypedContractEvent<
    OwnershipTransferredEvent.InputTuple,
    OwnershipTransferredEvent.OutputTuple,
    OwnershipTransferredEvent.OutputObject
  >;
  getEvent(
    key: "SafeDeployed"
  ): TypedContractEvent<
    SafeDeployedEvent.InputTuple,
    SafeDeployedEvent.OutputTuple,
    SafeDeployedEvent.OutputObject
  >;
  getEvent(
    key: "SafeOwnerSingletonUpgraded"
  ): TypedContractEvent<
    SafeOwnerSingletonUpgradedEvent.InputTuple,
    SafeOwnerSingletonUpgradedEvent.OutputTuple,
    SafeOwnerSingletonUpgradedEvent.OutputObject
  >;
  getEvent(
    key: "SingletonUpgraded"
  ): TypedContractEvent<
    SingletonUpgradedEvent.InputTuple,
    SingletonUpgradedEvent.OutputTuple,
    SingletonUpgradedEvent.OutputObject
  >;

  filters: {
    "OwnershipTransferred(address,address)": TypedContractEvent<
      OwnershipTransferredEvent.InputTuple,
      OwnershipTransferredEvent.OutputTuple,
      OwnershipTransferredEvent.OutputObject
    >;
    OwnershipTransferred: TypedContractEvent<
      OwnershipTransferredEvent.InputTuple,
      OwnershipTransferredEvent.OutputTuple,
      OwnershipTransferredEvent.OutputObject
    >;

    "SafeDeployed(address,address,bytes20)": TypedContractEvent<
      SafeDeployedEvent.InputTuple,
      SafeDeployedEvent.OutputTuple,
      SafeDeployedEvent.OutputObject
    >;
    SafeDeployed: TypedContractEvent<
      SafeDeployedEvent.InputTuple,
      SafeDeployedEvent.OutputTuple,
      SafeDeployedEvent.OutputObject
    >;

    "SafeOwnerSingletonUpgraded(address,address)": TypedContractEvent<
      SafeOwnerSingletonUpgradedEvent.InputTuple,
      SafeOwnerSingletonUpgradedEvent.OutputTuple,
      SafeOwnerSingletonUpgradedEvent.OutputObject
    >;
    SafeOwnerSingletonUpgraded: TypedContractEvent<
      SafeOwnerSingletonUpgradedEvent.InputTuple,
      SafeOwnerSingletonUpgradedEvent.OutputTuple,
      SafeOwnerSingletonUpgradedEvent.OutputObject
    >;

    "SingletonUpgraded(address,address)": TypedContractEvent<
      SingletonUpgradedEvent.InputTuple,
      SingletonUpgradedEvent.OutputTuple,
      SingletonUpgradedEvent.OutputObject
    >;
    SingletonUpgraded: TypedContractEvent<
      SingletonUpgradedEvent.InputTuple,
      SingletonUpgradedEvent.OutputTuple,
      SingletonUpgradedEvent.OutputObject
    >;
  };
}
