// @mezo-org/orangekit-contracts/typechain/contracts/DebugDeploy.ts
/* Autogenerated file. Do not edit manually. */
/* tslint:disable */
/* eslint-disable */
import type {
  BaseContract,
  BytesLike,
  FunctionFragment,
  Result,
  Interface,
  AddressLike,
  ContractRunner,
  ContractMethod,
  Listener,
} from "ethers";
import type {
  TypedContractEvent,
  TypedDeferredTopicFilter,
  TypedEventLog,
  TypedListener,
  TypedContractMethod,
} from "../common";

export interface DebugDeployInterface extends Interface {
  getFunction(
    nameOrSignature: "SALT" | "deployProxy" | "isContract"
  ): FunctionFragment;

  encodeFunctionData(functionFragment: "SALT", values?: undefined): string;
  encodeFunctionData(
    functionFragment: "deployProxy",
    values: [AddressLike]
  ): string;
  encodeFunctionData(
    functionFragment: "isContract",
    values: [AddressLike]
  ): string;

  decodeFunctionResult(functionFragment: "SALT", data: BytesLike): Result;
  decodeFunctionResult(
    functionFragment: "deployProxy",
    data: BytesLike
  ): Result;
  decodeFunctionResult(functionFragment: "isContract", data: BytesLike): Result;
}

export interface DebugDeploy extends BaseContract {
  connect(runner?: ContractRunner | null): DebugDeploy;
  waitForDeployment(): Promise<this>;

  interface: DebugDeployInterface;

  queryFilter<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;
  queryFilter<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    fromBlockOrBlockhash?: string | number | undefined,
    toBlock?: string | number | undefined
  ): Promise<Array<TypedEventLog<TCEvent>>>;

  on<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  on<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  once<TCEvent extends TypedContractEvent>(
    event: TCEvent,
    listener: TypedListener<TCEvent>
  ): Promise<this>;
  once<TCEvent extends TypedContractEvent>(
    filter: TypedDeferredTopicFilter<TCEvent>,
    listener: TypedListener<TCEvent>
  ): Promise<this>;

  listeners<TCEvent extends TypedContractEvent>(
    event: TCEvent
  ): Promise<Array<TypedListener<TCEvent>>>;
  listeners(eventName?: string): Promise<Array<Listener>>;
  removeAllListeners<TCEvent extends TypedContractEvent>(
    event?: TCEvent
  ): Promise<this>;

  SALT: TypedContractMethod<[], [string], "view">;

  deployProxy: TypedContractMethod<
    [singleton: AddressLike],
    [string],
    "nonpayable"
  >;

  isContract: TypedContractMethod<[account: AddressLike], [boolean], "view">;

  getFunction<T extends ContractMethod = ContractMethod>(
    key: string | FunctionFragment
  ): T;

  getFunction(
    nameOrSignature: "SALT"
  ): TypedContractMethod<[], [string], "view">;
  getFunction(
    nameOrSignature: "deployProxy"
  ): TypedContractMethod<[singleton: AddressLike], [string], "nonpayable">;
  getFunction(
    nameOrSignature: "isContract"
  ): TypedContractMethod<[account: AddressLike], [boolean], "view">;

  filters: {};
}
